codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -600,7 +604,41 @@ protected void onConnectionError(ChannelHandlerContext ctx, Throwable cause, Htt
      */
     protected void onStreamError(ChannelHandlerContext ctx, @SuppressWarnings("unused") Throwable cause,
                                  StreamException http2Ex) {
-        resetStream(ctx, http2Ex.streamId(), http2Ex.error().code(), ctx.newPromise());
+
+        int streamId = http2Ex.streamId();
+
+        //if this is caused by reading headers that are too large, send a header with status 431
+        if (http2Ex instanceof Http2Exception.HeaderListSizeException && connection().isServer()) {
+            final Http2Stream stream = connection().stream(streamId);
+
+            if (((Http2Exception.HeaderListSizeException) http2Ex).duringDecode()) {
```

Comment:
move this into the previous conditional for less branch statements and less indentation:

```java
if (http2Ex instanceof Http2Exception.HeaderListSizeException &&
    ((Http2Exception.HeaderListSizeException) http2Ex).duringDecode() &&
    connection().isServer()) {
..
}
```